l1-contracts/contracts/bridge/L1SharedBridge.sol

```diff
@@ -0,0 +1,637 @@
+    /// sent to the aliased `msg.sender` address.
+    /// @dev The address aliasing of L1 contracts as refund recipient on L2 is necessary to guarantee that the funds
+    /// are controllable through the Mailbox, since the Mailbox applies address aliasing to the from address for the
+    /// L2 tx if the L1 msg.sender is a contract. Without address aliasing for L1 contracts as refund recipients they
+    /// would not be able to make proper L2 tx requests through the Mailbox to use or withdraw the funds from L2, and
+    /// the funds would be lost.
+    /// @return l2TxHash The L2 transaction hash of deposit finalization.
+    function depositLegacyErc20Bridge(
+        address _prevMsgSender,
+        address _l2Receiver,
+        address _l1Token,
+        uint256 _amount,
+        uint256 _l2TxGasLimit,
+        uint256 _l2TxGasPerPubdataByte,
+        address _refundRecipient
+    ) external payable override onlyLegacyBridge nonReentrant returns (bytes32 l2TxHash) {
+        require(l2BridgeAddress[ERA_CHAIN_ID] != address(0), "ShB b. n dep");
+        require(_l1Token != l1WethAddress, "ShB: WETH deposit not supported 2");
+
+        // Note that funds have been transferred to this contract in the legacy ERC20 bridge.
+        if (!hyperbridgingEnabled[ERA_CHAIN_ID]) {
+            chainBalance[ERA_CHAIN_ID][_l1Token] += _amount;
+        }
+
+        bytes memory l2TxCalldata = _getDepositL2Calldata(_prevMsgSender, _l2Receiver, _l1Token, _amount);
+
+        {
+            // If the refund recipient is not specified, the refund will be sent to the sender of the transaction.
+            // Otherwise, the refund will be sent to the specified address.
+            // If the recipient is a contract on L1, the address alias will be applied.
+            address refundRecipient = _refundRecipient;
+            if (_refundRecipient == address(0)) {
+                refundRecipient = _prevMsgSender != tx.origin
+                    ? AddressAliasHelper.applyL1ToL2Alias(_prevMsgSender)
+                    : _prevMsgSender;
+            }
+
+            L2TransactionRequestDirect memory request = L2TransactionRequestDirect({
+                chainId: ERA_CHAIN_ID,
+                l2Contract: l2BridgeAddress[ERA_CHAIN_ID],
+                mintValue: msg.value, // l2 gas + l2 msg.Value the bridgehub will withdraw the mintValue from the base token bridge for gas
+                l2Value: 0, // L2 msg.value, this contract doesn't support base token deposits or wrapping functionality, for direct deposits use bridgehub
+                l2Calldata: l2TxCalldata,
+                l2GasLimit: _l2TxGasLimit,
+                l2GasPerPubdataByteLimit: _l2TxGasPerPubdataByte,
+                factoryDeps: new bytes[](0),
+                refundRecipient: refundRecipient
+            });
+            l2TxHash = bridgehub.requestL2TransactionDirect{value: msg.value}(request);
```

Comment:
I don't really get this.

For this line, the `L1SharedBridge` contract calls the `requestL2TransactionDirect` function on the `BridgeHub` contract, so `address(L1SharedBridge)` is the `msg.sender`.

Ok.

Now, `bridgehub.requestL2TransactionDirect` calls back into the `sharedBridge` contract via the `sharedBridge.bridgehubDepositBaseToken` function with the `prevSender` parameter set to `msg.sender` which would be `address(sharedBridge)`... which means, the following line:

```
uint256 amount = _depositFunds(_prevMsgSender, IERC20(_l1Token), _amount); // note if _prevMsgSender is this contract, this will return 0. This does not happen.
```

will either fail (since many tokens revert on self-transfer and also since it does not have self approval) or it will return 0, which means it will fail in the next `require` statement.

However, this case does not arrive if the _l1Token happens to be ETH.

Something is definitely off here. This is inconsistent behaviour.

Reply:
This is the ` depositLegacyErc20Bridge` function, so it is only for Era, with Eth as base token. So in the `bridgehubDepositBaseToken` we will have the Eth case, so it is ok, no? In particular, we will never call `_depositFunds`.

Reply:
Note that the `l1Token` variable will be different in different functions, in the `depositLegacyErc20Bridge` it will be some erc20, but in `bridgehubDepositBaseToken` it is the basetoken, so ether.